cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,154 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type contextKey string
+
+const (
+	OtelMeterCtxKey  = contextKey("otel.meter")
+	OtelTracerCtxKey = contextKey("otel.tracer")
+)
+
+// BaseMetricAttributes returns an attribute.Set containing attributes to attach to metrics/traces
+func BaseMetricAttributes(cmd *cobra.Command) attribute.Set {
+	attrList := []attribute.KeyValue{
+		attribute.String("command.name", getCommandName(cmd)),
+	}
+	return attribute.NewSet(attrList...)
+}
+
+// InstrumentCobraCommands wraps all cobra commands' RunE funcs to set a command duration metric using otel.
+//
+// Note: this should be the last func to wrap/modify the PersistentRunE/RunE funcs before command execution.
+//
+// can also be used for spans!
+func InstrumentCobraCommands(cmd *cobra.Command) {
+	// If PersistentPreRunE is nil, make it execute PersistentPreRun and return nil by default
+	ogPersistentPreRunE := cmd.PersistentPreRunE
+	if ogPersistentPreRunE == nil {
+		ogPersistentPreRun := cmd.PersistentPreRun
+		ogPersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+			ogPersistentPreRun(cmd, args)
+			return nil
+		}
+		cmd.PersistentPreRun = nil
+	}
+
+	// wrap RunE in PersistentPreRunE so that this operation gets executed on all children commands
+	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		// If RunE is nil, make it execute Run and return nil by default
+		ogRunE := cmd.RunE
+		if ogRunE == nil {
+			ogRun := cmd.Run
+			ogRunE = func(cmd *cobra.Command, args []string) error {
+				ogRun(cmd, args)
+				return nil
+			}
+			cmd.Run = nil
+		}
+		cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			// start the timer as the first step of every cobra command
+			stopCobraCmdTimer := startCobraCommandTimer(cmd)
+			cmdErr := ogRunE(cmd, args)
+			stopCobraCmdTimer(cmdErr)
+			return cmdErr
+		}
+
+		return ogPersistentPreRunE(cmd, args)
+	}
+}
+
+func startCobraCommandTimer(cmd *cobra.Command) func(err error) {
+	ctx := cmd.Context()
+	meter := GetDefaultMeter(otel.GetMeterProvider())
```

Comment:
nitpick: Can you modify `InstrumentCobraCommands` to take the `MeterProvider` as an input and use that instead of the global function?

I'm not sure of what the best way is, but the way I've been treating global state like `otel.GetMeterProvider` and `otel.SetMeterProvider` is they're only safe in code that is either in a `cmd` package or directly invoked only from a `cmd` package and everything else should take it as a parameter so an alternative method can be used.

Since the `cli/command` package is used as an API by plugins too, I don't think we want to force plugins to use `otel.SetMeterProvider`, but I think it's good if the plugins use `otel.SetMeterProvider` because it's useful and why not?

For this situation, you may not even need to use `otel.SetMeterProvider` anymore because I think they're used in the same scope anyway.

Reply:
Alright, I passed the MeterProvider along for now but still used otel.SetMeterProvider